### DISCUSSION TITLE
ModuleNotFoundError Cyptodome and Crypto

### DISCUSSION THREAD

**Discussion by LoveBooks** [noDRM/DeDRM_tools#187] [Q&A] [answered]
Using calibre for LInux verion 6.7. Using 10.0.3 of DeDRM from the 14th of July 2022. I go into calibre preferences, modify the behaviour of the DeDRM extension. I choose "Adobe Digital Editions ebooks". I choose to add an "Adobe Digital Editions Keys". I receive the following message:

DeDRM v10.0.3: Python3 exec found as wine py.exe -3
DeDRM v10.0.3: Running adobekey.py under Wine
Traceback (most recent call last):
  File "Z:\home\<username>\.config\calibre\plugins\DeDRM\libraryfiles\adobekey.py", line 130, in <module>
    from Cryptodome.Cipher import AES
ModuleNotFoundError: No module named 'Cryptodome'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "Z:\home\<username>\.config\calibre\plugins\DeDRM\libraryfiles\adobekey.py", line 132, in <module>
    from Crypto.Cipher import AES
ModuleNotFoundError: No module named 'Crypto'
DeDRM v10.0.3: Wine subprocess call error: 1
DeDRM v10.0.3: Found and decrypted 0 key files

I have installed Adobe Digital Editions version 4.5.11 underneath wine. That appears to run without a problem. I can open an .acsm file and download content from Adobe. 
Underneath the "Adobe Digital Editions Keys", if I put the WINEPREFIX: as /home/<username>/.wine, this does not help the situation.

**Comment by ElleKayEm** [comment, marked as answer]
I have zero experience with Linux, but it looks like you need to install PyCryptodome under Wine.

Also don't add the Adobe Key from within the DeDRM customization -- there's a bug.  Just add an Adobe DRM book to calibre and let DeDRM automatically retrieve the key.

**Reply by LoveBooks** [reply]
Hi, thanks for that info. That really put me on the right pathway. Initially I went under Linux (Ubuntu) and was installing pycryptodome under linux.Then I watched the running of calibre from a terminal and I saw the following:

DeDRM v10.0.3: Python3 exec found as wine py.exe -3

So, obviously, I need to install pycryptodome under wine (as a Windows precompiled binary).

So I went here, https://pycryptodome-master.readthedocs.io/en/latest/src/installation.html and I found the section on "Windows (pre-compiled)". It would be way too complicated for me to compile pycryptodome as I would have to install Visual Studio 2019 under wine on linux. Way too hard. That idea was outlined here: (under the section "Windows (from sources)")

https://pycryptodome.readthedocs.io/en/v3.15.0/src/installation.html

So I installed wheel. Remember, all this needs to be done under wine, not the general linux system.

<username>@<username>-desktop:~/.wine/drive_c/users/\<username\>/Local Settings/Application Data/Programs/Python/Python310-32/Scripts$ wine ./pip.exe install wheel
Collecting wheel
  Downloading wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel
  WARNING: The script wheel.exe is installed in 'C:\users\\<username\>\Local Settings\Application Data\Programs\Python\Python310-32\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed wheel-0.37.1

Now, I install pycryptodomex from pip.exe

<username>@<username>-desktop:~/.wine/drive_c/users/\<username\>/Local Settings/Application Data/Programs/Python/Python310-32/Scripts$ wine ./pip.exe install pycryptodomex
Collecting pycryptodomex
  Downloading pycryptodomex-3.15.0-cp35-abi3-win32.whl (1.9 MB)
     ---------------------------------------- 1.9/1.9 MB 2.4 MB/s eta 0:00:00
Installing collected packages: pycryptodomex
Successfully installed pycryptodomex-3.15.0

Yes, and as you say if I add the Adobe Key from within the DeDRM customization, it causes a big mess. I went in and deleted the key that I had added there. Now the the big win with decryption. As you say, I simply open the the encrypted pdf from the "My Digital Editions" directory of the <home>/Documents directory.

It worked. No worries. Here is the output when I opened the encrypted pdf. Works great.

DeDRM v10.0.3: Trying to decrypt Schizophrenia.pdf
DeDRM v10.0.3: Schizophrenia.pdf is a PDF ebook with encryption EBX_HANDLER
DeDRM v10.0.3: Schizophrenia.pdf is a PDF ebook (EBX) for UUID 05bd6bc9-4b29-46f5-b8b6-040e7854b379
DeDRM v10.0.3: Looking for new default Adobe Digital Editions Keys after 0.2 seconds
DeDRM v10.0.3: Python3 exec found as wine py.exe -3
DeDRM v10.0.3: Running adobekey.py under Wine
adobekey.py v7.4
Copyright © 2009-2020 i♥cabbages, Apprentice Harper et al.
Found 1 keys
Saved a key to Z:\home\\<username\>\\.config\calibre\plugins\DeDRM\libraryfiles\winekeysdir\adobekey1_uuid_05bd6bc9-4b29-46f5-b8b6-040e7854b379_AdobeID_\<username\>@gmail.com.
der
DeDRM v10.0.3: Found and decrypted 1 key file
DeDRM v10.0.3: Trying a new default key
DeDRM v10.0.3: Saving a new default key
DeDRM v10.0.3: Saved a new default key after 4.5 seconds
DeDRM v10.0.3: Finished after 4.5 seconds

**Reply by ElleKayEm** [reply]
Thanks for reporting back on how you got it working.
